File: 01_Language/01_Functions/js/array_walk.js
function array_walk(arr, callback) {
    if (!(typeof callback === 'function')) {
        return false
    }

    for (let i in arr) {
        if (arguments.length > 2) {
            callback(arr[i], i, arr, arguments[2])
        } else {
            callback(arr[i], i, arr)
        }
    }
    return true;
}

let fruits = {"d": "lemon", "a": "orange", "b": "banana", "c": "apple"}

function test_alter(item1, key, arr, prefix) {
    arr[key] = `${prefix}:${item1}`;
}

function test_print(item2, key) {
    console.log(`${key}: ${item2}`)
}

console.log("Before ...:")
array_walk(fruits, test_print);
array_walk(fruits, test_alter, 'fruit');
console.log("... and after:")
array_walk(fruits, test_print);

File: 01_Language/01_Functions/python/array_walk.py
# coding: utf-8

def array_walk(d, cb, *args):
    for k, v in d.items():
        cb(v, k, d, *args)


if __name__ == '__main__':
    fruits = {"d": "lemon", "a": "orange", "b": "banana", "c": "apple"}


    def test_alter(item1, key, arr, prefix):
        arr[key] = prefix + ":" + item1


    def test_print(item2, key, arr):
        print(key + ":" + item2)


    print("Before ...:")
    array_walk(fruits, test_print)
    array_walk(fruits, test_alter, 'fruit')
    print("... and after:")
    array_walk(fruits, test_print)
